[NCCL] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43232 [NCCL] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors**
* #41054 [NCCL] Destructor Blocks on WorkNCCL Completion
* #41053 [NCCL] WorkNCCL Helper Functions
* #41052 [NCCL] Abort Errored and Timed Out NCCL Communicators from Watchdog Thread
* #41051 [NCCL] Use cudaEventQuery to Poll for GPU operation errors
* #41050 [NCCL] Timeout Loop Thread for Async Error Handling

**This Commit:**
Here we introduce some experimental lock optimizations. This essentially checking for workNCCL completion and `handleNCCLGuard` so that the lock is acquired once instead of twice. The performance implications of this are still being measured. The first 5 PR's in this stack work perfectly well without this optimization.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D22929042](https://our.internmc.facebook.com/intern/diff/D22929042/)

osalpekar committed:
[WIP] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors

This method avoids the expensive serialization from adding a callback
and instead polls the CUDA event to check for completion. It then performs the
same error handling to throw an exception from the workCleanupThread.

Differential Revision: [D22929042](https://our.internmc.facebook.com/intern/diff/D22929042/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors"


This method avoids the expensive serialization from adding a callback
and instead polls the CUDA event to check for completion. It then performs the
same error handling to throw an exception from the workCleanupThread.

Differential Revision: [D22929042](https://our.internmc.facebook.com/intern/diff/D22929042/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors"


**This Commit:**
Here we introduce some experimental lock optimizations. This essentially checking for workNCCL completion and `handleNCCLGuard` so that the lock is acquired once instead of twice. The performance implications of this are still being measured. The first 5 PR's in this stack work perfectly well without this optimization.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D22929042](https://our.internmc.facebook.com/intern/diff/D22929042/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors"


**This Commit:**
Here we introduce some experimental lock optimizations. This essentially checking for workNCCL completion and `handleNCCLGuard` so that the lock is acquired once instead of twice. The performance implications of this are still being measured. The first 5 PR's in this stack work perfectly well without this optimization.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D22929042](https://our.internmc.facebook.com/intern/diff/D22929042/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors"


**This Commit:**
Here we introduce some experimental lock optimizations. This essentially checking for workNCCL completion and `handleNCCLGuard` so that the lock is acquired once instead of twice. The performance implications of this are still being measured. The first 5 PR's in this stack work perfectly well without this optimization.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D22929042](https://our.internmc.facebook.com/intern/diff/D22929042/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors"


**This Commit:**
Here we introduce some experimental lock optimizations. This essentially checking for workNCCL completion and `handleNCCLGuard` so that the lock is acquired once instead of twice. The performance implications of this are still being measured. The first 5 PR's in this stack work perfectly well without this optimization.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D22929042](https://our.internmc.facebook.com/intern/diff/D22929042/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -627,7 +640,7 @@ void ProcessGroupNCCL::workCleanupLoop() {
+      if (work->isCompletedAndThrowException()) {
         // Handle Exceptions on failed GPU operations and remove completed
         // workNCCL objects from work vector.
         if (work->finishedGPUExecution()) {
```
We probably need to remove this if we're already checking above?

osalpekar replied:
Right, this is removed in the latest version.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -627,7 +640,7 @@ void ProcessGroupNCCL::workCleanupLoop() {
          /* no increment*/) {
       auto& work = *it;
-      if (work->isCompleted()) {
+      if (work->isCompletedAndThrowException()) {
```
I don't think its safe to throw an exception from here since we'll kill the thread and crash the entire process.

jiayisuse replied:
We revised the solution from guard function registered by cudaStreamAddCallback, to throw exception in this cleanup thread. I guess here is intended to crash the entire process to let elastic launcher catch and resume.

osalpekar committed:
Update on "[NCCL] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors"


**This Commit:**
Here we introduce some experimental lock optimizations. This essentially checking for workNCCL completion and `handleNCCLGuard` so that the lock is acquired once instead of twice. The performance implications of this are still being measured. The first 5 PR's in this stack work perfectly well without this optimization.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D22929042](https://our.internmc.facebook.com/intern/diff/D22929042/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors"


**This Commit:**
Here we introduce some experimental lock optimizations. This essentially checking for workNCCL completion and `handleNCCLGuard` so that the lock is acquired once instead of twice. The performance implications of this are still being measured. The first 5 PR's in this stack work perfectly well without this optimization.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D22929042](https://our.internmc.facebook.com/intern/diff/D22929042/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] use cudaEventQuery instead of cudaStreamAddCallback to catch NCCL errors"


**This Commit:**
Here we introduce some experimental lock optimizations. This essentially checking for workNCCL completion and `handleNCCLGuard` so that the lock is acquired once instead of twice. The performance implications of this are still being measured. The first 5 PR's in this stack work perfectly well without this optimization.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D22929042](https://our.internmc.facebook.com/intern/diff/D22929042/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -306,6 +306,19 @@ void ProcessGroupNCCL::WorkNCCL::checkAndThrowException() {
+  return false;
+}
+
 void ProcessGroupNCCL::WorkNCCL::handleNCCLGuard() {
```
We no longer need this function?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -632,10 +645,9 @@ void ProcessGroupNCCL::workCleanupLoop() {
-        work->handleNCCLGuard();
+      // Handle Exceptions on failed GPU operations and remove completed
+      // workNCCL objects from work vector.
+      if (work->isCompletedAndThrowException()) {
```
It would be nice to add a LOG(ERROR) before we throw an exception mentioning the following: "Some NCCL operations have timed out/failed and due to the async nature of CUDA kernels subsequent GPU operations might run on corrupted/incomplete data. To avoid this inconsistency, we're taking the entire process down."

osalpekar replied:
Created https://github.com/pytorch/pytorch/pull/44988 with this change

pritamdamania87 reviewed (commented):
Looks good overall, have a few minor comments inline.

Can we also have another PR on top of this to remove the busy waiting in `wait()` and replace it with `cudaEventSynchronize`?
